### ISSUE TITLE
"for_each" not working in a scenario from terraform docs

### ISSUE THREAD

**Issue by michal-rybinski** [hashicorp/terraform-provider-aws#16437]
I've encountered strange behaviour and am not sure why.
I am writing a module that spins up a VPC with network plumbing and along those things are route table and route table association where I use for_each to iterate through subnets in vpc:

```
data "aws_subnet_ids" "mg_vpc_subnet_ids" {
  vpc_id = aws_vpc.mg_vpc.id
}

resource "aws_route_table_association" "mg_route_table_association" {
  for_each       = data.aws_subnet_ids.mg_vpc_subnet_ids.ids
  subnet_id      = each.value
  route_table_id = aws_route_table.mg_route_table.id
}
```

what is quite similar to the example which is in terraform docs(https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/subnet_ids):

```
data "aws_subnet_ids" "private" {
  vpc_id = var.vpc_id

  tags = {
    Tier = "Private"
  }
}

resource "aws_instance" "app" {
  for_each      = data.aws_subnet_ids.example.ids
  ami           = var.ami
  instance_type = "t2.micro"
  subnet_id     = each.value
}
```

When trying to plan my code generates the following error:

```
Error: Invalid for_each argument

  on ../../../modules/MediaGuardVPC/main.tf line 73, in resource "aws_route_table_association" "mg_route_table_association":
  73:   for_each       = data.aws_subnet_ids.mg_vpc_subnet_ids.ids

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.
```

I am a bit at a loss of why it is not able to calculate it.

Terraform v0.13.5
hashicorp/aws: version = "~> 3.17.0"
MacOS 11

Thanks,
Michal

**Comment by michal-rybinski** [comment]
not sure how to change labels added automatically by bots... I am guessing this should not have the service/ec2 label on it :/

**Comment by bflad** [comment]
Hi @michal-rybinski 👋  Thank you for filing this and sorry you ran into trouble here.

The resources and data sources in Terraform AWS Provider codebase have little control over this particular behavior since it relates to the operation graph handling that is performed within the Terraform CLI codebase. Your best bet is to create an issue in that repository so that the people familiar with that handling can potentially triage this issue: https://github.com/hashicorp/terraform/issues

My best guess is that this is likely fixed in the upcoming Terraform 0.14 release since it better integrates data sources into the graph. The upstream maintainers would know best though. There is the Terraform 0.14.0-rc1 prerelease available if you have a development or test environment where you could potentially try that out.
